File: backend/app/schemas/api_envelope.py
from __future__ import annotations

from typing import Any, Generic, Optional, TypeVar

from pydantic import BaseModel, Field

T = TypeVar("T")


class ApiMeta(BaseModel):
    request_id: str = Field(default="-")


class ApiError(BaseModel):
    code: str
    message: str
    details: Optional[Any] = None


class ApiEnvelope(BaseModel, Generic[T]):
    data: Optional[T] = None
    meta: ApiMeta
    error: Optional[ApiError] = None

File: frontend/src/lib/api-envelope.ts
export interface ApiEnvelopeError {
  code: string;
  message: string;
  details?: unknown;
}

export interface ApiEnvelope<T = unknown> {
  data: T | null;
  meta: {
    request_id?: string;
    [key: string]: unknown;
  };
  error: ApiEnvelopeError | null;
}

export function isApiEnvelopePayload(payload: unknown): payload is ApiEnvelope {
  if (!payload || typeof payload !== 'object') return false;
  const record = payload as Record<string, unknown>;
  return 'data' in record && 'meta' in record && 'error' in record;
}

export function unwrapApiEnvelopeData<T = unknown>(payload: unknown): T | unknown {
  if (!isApiEnvelopePayload(payload)) return payload;
  return payload.data;
}

export function normalizeApiEnvelopeError(payload: unknown): {
  detail: unknown;
  message: string;
  code: string;
} | null {
  if (!isApiEnvelopePayload(payload) || !payload.error) {
    return null;
  }
  const details = payload.error.details;
  return {
    detail: details ?? payload.error.message,
    message: payload.error.message,
    code: payload.error.code,
  };
}

